### ISSUE TITLE
[Bug] 17x E2E slowdown in SGLang when `return_hidden_states=True`

### ISSUE THREAD

**Issue by Sinestro38** [sgl-project/sglang#8003]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

Why is SGLang 10x+ slower when asked to return the hidden_states (last layer only)? This seems almost unusable. The slowdown is only 4x for prompts with <100 tok and exceeds 15x with larger prompts (2k+ tokens).

I used some nsys traces and got to the bottom of this discrepancy in performance between when return_hidden_states=True vs False.
### Problem (not gpu kernel choice):
- The data traffic is tiny, but copied the wrong way. 36 Device→Host copies, 50.9 MB total, 0.97 ms GPU time (which is good), but wall clock goes up by 4.7s (!!).
- GPU stays idle 90% of time (encode-heavy workloads)
### Root cause:
- hidden states are copied synchronously for every chunk/token and converted to python lists instead of at the end.

[with_hidden_states.txt](https://github.com/user-attachments/files/21208286/with_hidden_states.txt)
[no_hidden_states.txt](https://github.com/user-attachments/files/21208285/no_hidden_states.txt)

Attached are the nsys traces, showing how turning this on in the SGLang engine increases latency by 17x for the exact same benchmark run.

### Reproduction

```
"""
Usage:
python sgl_test.py
"""

import torch
import time, nvtx
import sglang as sgl

BATCH_SIZE = 8
DUMMY_TOKEN = 100
PROMPT_LEN = 2000

def benchmark_generate(enable_hs: bool):
    # Create an LLM.
    llm = sgl.Engine(
        model_path="Qwen/Qwen3-8B",
    )

    sampling_params = {
        "temperature": 0.8,
        "top_p": 0.95,
        "max_new_tokens": 10,
        "ignore_eos": True,
    }

    warmup_prompts = [
		"Once upon a time in a distant land,",
		"A healthy breakfast usually includes",
		"The fastest animal on Earth is",
		"In the year 2050, humans might",
	]
    _ = llm.generate(warmup_prompts, sampling_params=sampling_params, return_hidden_states=enable_hs)
	
    prompt_ids = [[DUMMY_TOKEN for i in range(PROMPT_LEN)] for b in range(BATCH_SIZE)]
    with nvtx.annotate(f"{'with ' if enable_hs else 'no '}hidden states", color="green"):
        start_t = time.perf_counter()
        outputs = llm.generate(
            input_ids=prompt_ids, sampling_params=sampling_params, return_hidden_states=enable_hs
            )
        end_t = time.perf_counter()
    print(f"Time taken to generate (return_hidden_states={enable_hs}): {(end_t - start_t):.2f} s")

    llm.shutdown()

def main():
    benchmark_generate(enable_hs=True)
    benchmark_generate(enable_hs=False)

if __name__ == "__main__":
    main()
```

### Output showing massive slowdown (worse than huggingface)
```
bash-5.1$ python sgl_test.py 
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:00<00:00,  7.14it/s]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:00<00:00,  2.40it/s]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:01<00:00,  1.93it/s]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:02<00:00,  1.76it/s]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:02<00:00,  1.96it/s]

Capturing batches (avail_mem=7.81 GB): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23/23 [00:02<00:00,  7.67it/s]
Capturing batches (avail_mem=6.80 GB): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23/23 [00:03<00:00,  7.59it/s]
*** Time taken to generate (return_hidden_states=True): 11.38 s ***
/my-conda-envs/lib64/python3.11/multiprocessing/resource_tracker.py:123: UserWarning: resource_tracker: process died unexpectedly, relaunching.  Some resources might leak.
  warnings. Warn('resource tracker: process died unexpectedly, '
Loading safetensors checkpoint shards:   0% Completed | 0/4 [00:00<?, ?it/s]
Loading safetensors checkpoint shards:  25% Completed | 1/4 [00:00<00:00,  7.14it/s]
Loading safetensors checkpoint shards:  50% Completed | 2/4 [00:00<00:00,  2.41it/s]
Loading safetensors checkpoint shards:  75% Completed | 3/4 [00:01<00:00,  1.93it/s]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:02<00:00,  1.75it/s]
Loading safetensors checkpoint shards: 100% Completed | 4/4 [00:02<00:00,  1.96it/s]

Capturing batches (avail_mem=7.81 GB): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 23/23 [00:03<00:00,  7.65it/s]
*** Time taken to generate (return_hidden_states=False): 0.58 s ***
bash-5.1$ 
```

### Environment

```
Python: 3.11.12 (main, May 20 2025, 12:28:06) [GCC 12.1.0]
CUDA available: True
GPU 0: NVIDIA H100 NVL
GPU 0 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.3, V12.3.107
CUDA Driver Version: 570.133.20
PyTorch: 2.6.0.post3
sglang: 0.4.6.post5
sgl_kernel: 0.1.4
flashinfer_python: Module Not Found
triton: 3.1.0
transformers: 4.50.3
torchao: 0.8.0+git
numpy: 1.26.4
aiohttp: 3.8.5
fastapi: 0.109.2
hf_transfer: Module Not Found
huggingface_hub: 0.31.4
interegular: Module Not Found
modelscope: Module Not Found
orjson: Module Not Found
outlines: Module Not Found
packaging: 25.0
psutil: 5.9.8
pydantic: 2.11.7
python-multipart: Module Not Found
pyzmq: 27.0.0
uvicorn: Module Not Found
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.1.19
openai: 1.93.0
tiktoken: 0.9.0
anthropic: Module Not Found
litellm: Module Not Found
decord: Module Not Found
NVIDIA Topology: 
        GPU0    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      0-31    0               N/A

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

ulimit soft: 131072

```

**Comment by Sinestro38** [comment]
Cc: @Qiaolin-Yu , bumping this for visibility.

**Comment by Qiaolin-Yu** [comment]
Hi @Sinestro38 , thanks for reporting. May you open a PR to fix it?

**Comment by Sinestro38** [comment]
Hi @Qiaolin-Yu , thank you for your prompt response, given the nsys traces and the root cause I described (hidden states are copied synchronously for every chunk from D->H and GPU/scheduler is on hold until the copy is materialized as a python list), can you describe the files you think I'd need to modify to move the D->H hidden state copies to the end of the request?

**Comment by Sinestro38** [comment]
@Qiaolin-Yu Also updated the issue description to include the `stdout` once you run the repro script, demonstrating a `11.38s` E2E generate with hidden states enabled and `0.58s` E2E generate with it disabled, all things held constant. Thank you for your guidance and great contribution to this community!

**Comment by Qiaolin-Yu** [comment]
Maybe you can take a look at #3364 and find the related file? This is the PR which introduces this feature.

**Comment by Sinestro38** [comment]
Thanks! I'll look at this. Then perhaps I should cc @Jackmin801 into this as well since he pioneered this feature :)

**Comment by Jackmin801** [comment]
hrmm im not sure if u can delay the d2h copy to the end of the request. reason being cuda graphs reuse the same memory locations so you have to clone them out before the next cuda graph call. you could try to clone them d2d instead of d2h though. might be faster if it works

**Comment by Jackmin801** [comment]
Modifying scheduler.py should be what youre looking for https://github.com/sgl-project/sglang/pull/3364/files#diff-c3b8cc39d10c245933a25aa9c2fd6397f6b31ed8d85c0ecbb926c1f42afdd178

namely the 2 `cpu().clone()` lines
